Userland/Libraries/LibGfx/ImageFormats/CCITTDecoder.cpp

```diff
@@ -349,6 +361,152 @@ static ErrorOr<void> read_eol(BigEndianInputBitStream& bit_stream, Group3Options
     return {};
 }
 
+enum class Mode : u8 {
+    Pass,
+    Horizontal,
+    Vertical_0,
+    Vertical_R1,
+    Vertical_R2,
+    Vertical_R3,
+    Vertical_L1,
+    Vertical_L2,
+    Vertical_L3,
+};
+
+struct ModeCode {
+    u8 code_length {};
+    Mode mode {};
+    u8 code {};
+};
+
+// Table 4/T.4 – Two-dimensional code table
+constexpr Array node_codes = to_array<ModeCode>({
+    { 4, Mode::Pass, 0b0001 },
+    { 3, Mode::Horizontal, 0b001 },
+    { 1, Mode::Vertical_0, 0b1 },
+    { 3, Mode::Vertical_R1, 0b011 },
+    { 6, Mode::Vertical_R2, 0b000011 },
+    { 7, Mode::Vertical_R3, 0b0000011 },
+    { 3, Mode::Vertical_L1, 0b010 },
+    { 6, Mode::Vertical_L2, 0b000010 },
+    { 7, Mode::Vertical_L3, 0b0000010 },
+});
+
+ErrorOr<ModeCode> read_mode(BigEndianInputBitStream& input_bit_stream)
+{
+    u8 size {};
+    u16 potential_code {};
+    while (size < 7) {
+        potential_code <<= 1;
+        potential_code |= TRY(input_bit_stream.read_bit());
+        ++size;
+
+        if (auto const maybe_mode = get_code_from_table(node_codes, potential_code, size); maybe_mode.has_value())
+            return *maybe_mode;
+    }
+
+    return Error::from_string_literal("CCITTDecoder: Unable to find the correct mode");
+}
+
+enum class Search : u8 {
+    B1,
+    B2,
+};
+
+ErrorOr<void> decode_single_ccitt3_2d_block(BigEndianInputBitStream& input_bit_stream, BigEndianOutputBitStream& decoded_bits, u32 image_width, u32 image_height, Group3Options::UseFillBits use_fill_bits)
+{
+    ReferenceLine reference_line;
+    for (u32 i = 0; i < image_height; ++i) {
+        TRY(read_eol(input_bit_stream, use_fill_bits));
+        bool const next_is_1D = TRY(input_bit_stream.read_bit()) == 1;
+
+        if (next_is_1D) {
+            reference_line = TRY(decode_single_ccitt3_1d_line(input_bit_stream, decoded_bits, image_width));
+        } else {
+            ReferenceLine current_line {};
+            Color current_color { ccitt_white };
+            u32 column {};
+
+            auto const next_change_on_reference_line = [&](Search search = Search::B1) -> ErrorOr<Change> {
+                // 4.2.1.3.1 Definition of changing picture elements
+                Optional<Change> next_change {}; // This is referred as b1 in the spec
```

Comment:
"This is called b1 in the spec" or "This is referred to as b1 in the spec" I think